3-enrich/scala-common-enrich/src/main/scala/com.snowplowanalytics.snowplow.enrich/common/adapters/registry/RemoteAdapter.scala

```diff
@@ -0,0 +1,96 @@
+
+import com.snowplowanalytics.iglu.client.Resolver
+import com.snowplowanalytics.snowplow.enrich.common.loaders.CollectorPayload
+import com.softwaremill.sttp._
+import org.json4s.JsonDSL._
+import org.json4s.jackson.JsonMethods._
+import org.slf4j.LoggerFactory
+import scalaz.Scalaz._
+
+import scala.concurrent.duration.FiniteDuration
+import scala.util.{Failure, Success, Try}
+import scala.util.control.NonFatal
+
+/**
+ * An adapter for an enrichment that is handled by a remote webservice.
+ *
+ * @constructor create a new client to talk to the given remote webservice.
+ * @param remoteUrl the url of the remote webservice, e.g. http://localhost/myEnrichment
+ * @param timeout max duration of each connection attempt, and of the wait for each response from the remote
+ */
+class RemoteAdapter(val remoteUrl: String, val timeout: FiniteDuration) extends Adapter {
+
+  val bodyMissingErrorText = "missing payload body"
+
+  private lazy val log = LoggerFactory.getLogger(getClass)
+
+  val remoteUri = uri"$remoteUrl"
+  implicit val backend = HttpURLConnectionBackend(
+    options = SttpBackendOptions.connectionTimeout(timeout)
+  )
+
+  /**
+   * POST the given payload to the remote webservice,
+   * wait for it to respond with an Either[List[String], List[RawEvent] ],
+   * and return that as a ValidatedRawEvents
+   *
+   * @param payload The CollectorPaylod containing one or more
+   *        raw events as collected by a Snowplow collector
+   * @param resolver (implicit) The Iglu resolver used for
+   *        schema lookup and validation
+   * @return a Validation boxing either a NEL of RawEvents on
+   *         Success, or a NEL of Failure Strings
+   */
+  def toRawEvents(payload: CollectorPayload)(implicit resolver: Resolver): ValidatedRawEvents =
+    payload.body match {
+      case Some(body) if body.nonEmpty =>
+        val json = ("contentType" -> payload.contentType) ~
+          ("queryString" -> toMap(payload.querystring)) ~
+          ("body"        -> payload.body)
+
+        httpPost(remoteUrl, compact(render(json))) match {
+          case Left(errmsg) =>
+            errmsg.failNel
+          case Right(bodyAsString) =>
+            try {
+              val error = (parse(bodyAsString) \ "error").extract[String]
+              if (error != null) {
+                error.failNel
+              } else {
+                val events = (parse(bodyAsString) \ "events").extract[List[Map[String, String]]]
+                rawEventsListProcessor(events.map { event =>
+                  RawEvent(
+                    api         = payload.api,
+                    parameters  = event,
+                    contentType = payload.contentType,
+                    source      = payload.source,
+                    context     = payload.context
+                  ).success
+                })
+              }
+            } catch {
+              case NonFatal(e) => s"Unexpected response from remote $e".failNel
+            }
```

Comment:
This `try catch` wraps many lines and different things (error and success), it would be nice to have more fine-grained control over the processing of the response.

Reply:
I have added a few things below. but I can imagine that would be it.